File: math/matrix.cpp
#include "matrix.h"
#include <cassert>

namespace math
{
    Matrix::Matrix(const std::size_t rows, const std::size_t cols) : arr{std::make_unique<std::unique_ptr<double[]>[]>(rows)},
                                                                     rows{rows},
                                                                     cols{cols}
    {
        for (std::size_t i = 0; i < rows; i++)
        {
            auto oneDimArr = std::make_unique<double[]>(cols);
            arr[i] = std::move(oneDimArr);
        }
    }

    Matrix::Matrix(const Matrix &matrix) : arr{std::make_unique<std::unique_ptr<double[]>[]>(matrix.rows)},
                                           rows{matrix.rows},
                                           cols{matrix.cols}
    {
        for (std::size_t row = 0; row < rows; row++)
        {
            auto oneDimArr = std::make_unique<double[]>(cols);
            arr[row] = std::move(oneDimArr);
            for (std::size_t col = 0; col < cols; col++)
            {
                at(row, col) = matrix.at_r(row, col);
            }
        }
    }

	Matrix::Matrix(const Vector& vector) : arr{ std::make_unique<std::unique_ptr<double[]>[]>(vector.getSize()) },
		                                   rows{ vector.getSize() },
		                                   cols{ 1 }
	{
        for (std::size_t row = 0; row < rows; row++)
        {
            auto oneDimArr = std::make_unique<double[]>(cols);
            arr[row] = std::move(oneDimArr);
            at(row, 0) = vector.at_r(row);
        }
	}

    Matrix::Matrix(Matrix &&matrix) : arr{std::move(matrix.arr)},
                                      rows{matrix.rows},
                                      cols{matrix.cols}
    {
    }

    Matrix::Matrix() : arr{nullptr},
                       rows{0},
                       cols{0}
    {
    }

    Matrix operator*(const Matrix &matrix, const double value)
    {
        Matrix ret(matrix);

        for (std::size_t row = 0; row < matrix.rows; row++)
        {
            for (std::size_t col = 0; col < matrix.cols; col++)
            {
                ret.at(row, col) *= value;
            }
        }

        return Matrix(std::move(ret));
    }

    Matrix operator*(const double value, const Matrix &matrix)
    {
        return operator*(matrix, value);
    }

    Vector operator*(const Matrix &matrix, const Vector &vector)
    {
        assert(matrix.cols == vector.getSize());
        Vector ret(matrix.rows);

        for (std::size_t row = 0; row < matrix.rows; row++)
        {
            for (std::size_t col = 0; col < matrix.cols; col++)
            {
                ret.at(row) += vector.at_r(col) * matrix.at_r(row, col);
        }
    }
        return Vector(std::move(ret));
    }

    Matrix operator+(const Matrix& lhs, const Matrix& rhs)
    {
        assert(lhs.rows == rhs.rows && lhs.cols == rhs.cols);
        Matrix ret{ lhs };

        for (std::size_t row = 0; row < lhs.rows; row++)
        {
            for (std::size_t col = 0; col < lhs.cols; col++)
            {
                ret.at(row, col) += rhs.at_r(row, col);
            }
        }

        return Matrix(std::move(ret));
    }

    Matrix operator-(const Matrix &lhs, const Matrix &rhs)
    {
        assert(lhs.rows == rhs.rows && lhs.cols == rhs.cols);
        Matrix ret{lhs};

        for (std::size_t row = 0; row < lhs.rows; row++)
        {
            for (std::size_t col = 0; col < lhs.cols; col++)
            {
                ret.at(row, col) -= rhs.at_r(row, col);
            }
        }

        return Matrix(std::move(ret));
    }

    Matrix operator*(const Matrix &lhs, const Matrix &rhs)
    {
        assert(lhs.cols == rhs.rows);
        Matrix ret{lhs.rows, rhs.cols};

        auto getMultipliedValue = [&lhs, &rhs](const std::size_t row, const std::size_t col) {
            double val = 0.0;

            for (std::size_t i = 0; i < lhs.cols; i++)
            {
                val += (lhs.at_r(row, i) * rhs.at_r(i, col));
            }
            return val;
        };

        for (std::size_t row = 0; row < lhs.rows; row++)
        {
            for (std::size_t col = 0; col < rhs.cols; col++)
            {
                ret.at(row, col) = getMultipliedValue(row, col);
            }
        }
        return Matrix(std::move(ret));
    }

    std::ostream &operator<<(std::ostream &out, const Matrix &matrix)
    {
        for (std::size_t row = 0; row < matrix.rows; row++)
        {
            out << "[";

            if (matrix.cols > 0)
            {
                out << " " << matrix.at_r(row, 0);
            }

            for (std::size_t col = 1; col < matrix.cols; col++)
            {
                out << ", " << matrix.at_r(row, col);
            }

            out << " ]" << std::endl;
        }
        return out;
    }
} // namespace math

File: AK7_CSHARP/Algebra/Matrix.cs
using System;
using System.Diagnostics;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace AK7_CSHARP.Algebra
{
    public class Matrix
    {
        public static int ThreadNumber { get; set; }

        public int Rows { get; }

        public int Columns { get; }

        public double[][] Elems { get; set; }

        public double this[int row, int col]
        {
            get => Elems[row][col];
            set => Elems[row][col] = value;
        }

        public double[] this[int row]
        {
            get => Elems[row];
            set => Elems[row] = value;
        }

        public Matrix(int rows, int columns)
        {
            Rows = rows;
            Columns = columns;
            Elems = new double[Rows][];

            for (int i = 0; i < Rows; i++)
            {
                Elems[i] = new double[Columns];
            }
        }

        public Matrix(Matrix other)
        {
            Rows = other.Rows;
            Columns = other.Columns;
            Elems = new double[Rows][];

            for (int i = 0; i < Rows; i++)
            {
                Elems[i] = new double[Columns];
                Array.Copy(other[i], Elems[i], Columns);
            }
        }

        public Matrix(Vector vector) : this(vector.Size, 1)
        {
            for (int i = 0; i < vector.Size; i++)
            {
                Elems[i][0] = vector[i];
            }
        }

        public static Matrix operator *(double value, Matrix other)
        {
            Matrix ret = new Matrix(other);

            for (int row = 0; row < ret.Rows; row++)
            {
                for (int col = 0; col < ret.Columns; col++)
                {
                    ret[row, col] *= value;
                }
            }

            return ret;
        }

        public static Matrix operator *(Matrix other, double value)
        {
            return value * other;
        }

        public static Vector operator *(Matrix matrix, Vector vector)
        {
            Debug.Assert(matrix.Columns == vector.Size);

            Vector ret = new Vector(matrix.Rows);

            for (int row = 0; row < matrix.Rows; row++)
            {
                for (int col = 0; col < matrix.Columns; col++)
                {
                    ret[row] += vector[col] * matrix[row, col];
                }
            }

            return ret;
        }

        public static Matrix operator +(Matrix lhs, Matrix rhs)
        {
            Debug.Assert(lhs.Rows == rhs.Rows && lhs.Columns == rhs.Columns);

            Matrix ret = new Matrix(lhs);

            for (int row = 0; row < lhs.Rows; row++)
            {
                for (int col = 0; col < lhs.Columns; col++)
                {
                    ret[row, col] += rhs[row, col];
                }
            }

            return ret;
        }

        public static Matrix operator -(Matrix lhs, Matrix rhs)
        {
            Debug.Assert(lhs.Rows == rhs.Rows && lhs.Columns == rhs.Columns);

            Matrix ret = new Matrix(lhs);

            for (int row = 0; row < lhs.Rows; row++)
            {
                for (int col = 0; col < lhs.Columns; col++)
                {
                    ret[row, col] -= rhs[row, col];
                }
            }

            return ret;
        }
        
        public static Matrix operator *(Matrix lhs, Matrix rhs)
        {
            Debug.Assert(lhs.Columns == rhs.Rows);

            Matrix ret = new Matrix(lhs.Rows, rhs.Columns);

            double getMultipliedValue(int row, int col)
            {
                double val = 0.0;
                for (int i = 0; i < lhs.Columns; i++)
                {
                    val += lhs[row, i] * rhs[i, col];
                }
                return val;
            }

#if USE_PARALLEL
            Parallel.For(0, lhs.Rows, new ParallelOptions { MaxDegreeOfParallelism = ThreadNumber }, row =>
            {
                for (int col = 0; col < rhs.Columns; col++)
                {
                    ret[row, col] = getMultipliedValue(row, col);
                }
            });
#else
            for (int row = 0; row < lhs.Rows; row++)
            {
                for (int col = 0; col < rhs.Columns; col++)
                {
                    ret[row, col] = getMultipliedValue(row, col);
                }
            }
#endif

            return ret;
        }

        public static Matrix Identity(int size)
        {
            Matrix ret = new Matrix(size, size);

            for (int row = 0; row < size; row++)
            {
                ret[row, row] = 1.0;
            }

            return ret;
        }

        public static Matrix Transpose(Vector vector)
        {
            Matrix ret = new Matrix(1, vector.Size);

            Array.Copy(vector.Elems, ret[0], vector.Size);
            
            return ret;
        }

        public static Matrix Negate(Matrix matrix)
        {
            Matrix ret = new Matrix(matrix);
            
            for (int row = 0; row < ret.Rows; row++)
            {
                for (int col = 0; col < ret.Columns; col++)
                {
                    ret[row, col] = -ret[row, col];
                }
            }

            return ret;
        }

        public double Scalar()
        {
            Debug.Assert(Rows == 1 && Columns == 1);

            return Elems[0][0];
        }

        public override string ToString()
        {
            return string.Join("\n", Elems.Select(x => "[ " + string.Join(", ", x) + " ]"));
        }
    }
}
